activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb

```diff
@@ -1023,7 +1023,11 @@ def mismatched_foreign_key(message, sql:, binds:, connection_pool:)
         end
 
         def version_string(full_version_string)
-          full_version_string.match(/^(?:5\.5\.5-)?(\d+\.\d+\.\d+)/)[1]
+          if full_version_string && matches = full_version_string.match(/^(?:5\.5\.5-)?(\d+\.\d+\.\d+)/)
+            matches[1]
+          else
+            raise ActiveRecordError, "Unable to parse MySQL version from #{full_version_string.inspect}"
```

Comment:
I think ConnectionError would be better. I don't like `ActiveRecord::ActiveRecordError`, it's too generic. Especially when the problem is on the db side not on the AR side. The other version errors raise a `RuntimeError`, so I guess that's an option too.

Reply:
I can do `ConnectionError`. I don't like `RuntimeError` because it's the default `raise "whatever"` and makes it hard to rescue or make assumptions about.

Reply:
I went with [`DatabaseConnectionError`](https://github.com/rails/rails/blob/3d418c05519c8d90816ca08ae9a8ea70118c2d36/activerecord/lib/active_record/errors.rb#L85).